### PR TITLE
Improvements in QuorumFuture exception management

### PR DESCRIPTION
**Pull request by kspirov**
The client changes requre QuorumFuture to permit:

* collection of all exceptions (not only the last). We need it when
  there is existing value from the server,in order to collect the one
  with highest rank.

* fail fast exceptions that will cause the future to return result imediately
(DataOutrankedException is such example)

(cherry picked from commit d6c2eec)

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0b6096f**.

 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #516 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/yYJDSuxZOx52Pec3UYpQM7iGK2PNIgwT)

File: runtime/src/main/java/org/corfudb/runtime/view/QuorumFuturesFactory.java [hunk 1]
```diff
@@ -186,7 +208,10 @@ public boolean isConflict() {
             return conflict;
         }
 
-        public Throwable getThrowable() {return throwable;}
+        public Set<Throwable> getThrowables() {
```
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/view/QuorumFuturesFactory.java]: Add javadoc and explanation of why you need to return an immutable.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **308d815**.

*** 0.0% transaction FAILURE rate for 1 threads, CorfuMemoryNode mode
*** 0.0% transaction FAILURE rate for 5 threads, CorfuMemoryNode mode
*** 0.0% transaction FAILURE rate for 10 threads, CorfuMemoryNode mode
*** 0.0% transaction FAILURE rate for 1 threads, CorfuDiskNode mode
*** 0.03% transaction FAILURE rate for 5 threads, CorfuDiskNode mode
*** 0.04% transaction FAILURE rate for 10 threads, CorfuDiskNode mode
*** 100.0% transaction FAILURE rate for 1 threads, CorfuMemoryNode mode
*** 100.0% transaction FAILURE rate for 5 threads, CorfuMemoryNode mode
*** 100.0% transaction FAILURE rate for 10 threads, CorfuMemoryNode mode
*** 100.0% transaction FAILURE rate for 1 threads, CorfuDiskNode mode
*** 100.0% transaction FAILURE rate for 5 threads, CorfuDiskNode mode
*** 100.0% transaction FAILURE rate for 10 threads, CorfuDiskNode mode

An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #516 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/DBiTDcQkAeaiXiAlkJ3RHGlfcM9CEKmp)

**Comment by no2chem** [comment]
Thanks for the work on this, looks good and merging in.

**Pull request merged**